### PR TITLE
[RC 1] Creation and Insertion of Account Recovery Codes

### PR DESCRIPTION
**Pull request by kimlisa**
part of RFD 0029: https://github.com/gravitational/teleport/blob/master/rfd/0029-account-lifecycle.md
split of from the bigger PR for easier review: https://github.com/gravitational/teleport/pull/7405

#### Description
- Adds dice-ware library to produce recovery codes
- Implement creation and upsertion of recovery codes
- Converted `ChangePasswordWithToken` to grpc endpoint (kept http for fallback) and renamed it to `ChangeUserAuthentication`. This endpoint now returns recovery codes for user who meet the following criteria:
  1. cloud is enabled
  2. second factor is enabled
  3. username is a valid email format
- There are some methods related to `recovery attempts` that are not utilized until later PRs

### CODE REVIEW

**Review by codingllama — COMMENTED** [review]
Partial review, will wrap up tomorrow.

File: lib/defaults/defaults.go [hunk 1]
```diff
@@ -224,6 +224,10 @@ const (
 	// before a user account is locked for AccountLockInterval
 	MaxLoginAttempts int = 5
 
+	// MaxAccountRecoveryAttempts sets the max number of allowed failed recovery attempts
+	// before a user is locked from login and further recovery attempts for AccountLockInterval.
+	MaxAccountRecoveryAttempts = 3
```
Comment (codingllama) [inline lib/defaults/defaults.go:229]: Was this used in the PR? (Sorry, I can't recall seeing it before.)

File: lib/services/identity.go [hunk 2]
```diff
@@ -207,6 +207,21 @@ type Identity interface {
...
+	GetRecoveryCodes(ctx context.Context, user string) (*types.RecoveryCodesV1, error)
+
+	// CreateUserRecoveryAttempt logs user recovery attempt.
+	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt types.RecoveryAttempt) error
+
+	// GetUserRecoveryAttempts returns user recovery attempts.
+	GetUserRecoveryAttempts(ctx context.Context, user string) ([]types.RecoveryAttempt, error)
+
+	// DeleteUserRecoveryAttempts removes all recovery attempts of a user.
+	DeleteUserRecoveryAttempts(ctx context.Context, user string) error
```
Comment (codingllama) [inline lib/services/identity.go:223]: When would we use this one?

File: lib/services/local/users.go [hunk 3]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+// GetRecoveryCodes returns user's recovery codes.
+func (s *IdentityService) GetRecoveryCodes(ctx context.Context, user string) (*types.RecoveryCodesV1, error) {
+	if user == "" {
+		return nil, trace.BadParameter("missing parameter user")
+	}
+
+	item, err := s.Get(ctx, backend.Key(webPrefix, usersPrefix, user, recoveryCodesPrefix))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, trace.NotFound("recovery codes for user %q is not found", user)
```
Comment (codingllama) [inline lib/services/local/users.go:1070]: Is this necessary? Can't we trace.Wrap all cases?

File: lib/services/local/users.go [hunk 4]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+}
+
+// UpsertRecoveryCodes creates or updates user's account recovery codes.
+// Each recovery code are hashed before upsert.
+func (s *IdentityService) UpsertRecoveryCodes(ctx context.Context, user string, recovery types.RecoveryCodesV1) error {
+	if user == "" {
+		return trace.BadParameter("missing parameter user")
+	}
+
+	if err := recovery.CheckAndSetDefaults(); err != nil {
```
Comment (codingllama) [inline lib/services/local/users.go:1090]: Should we check `recovery == nil` first?

File: lib/services/local/users.go [hunk 5]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+	return trace.Wrap(err)
+}
+
+// CreateUserRecoveryAttempt creates new user recovery attempt.
+func (s *IdentityService) CreateUserRecoveryAttempt(ctx context.Context, user string, attempt types.RecoveryAttempt) error {
+	if user == "" {
+		return trace.BadParameter("missing parameter user")
+	}
+
+	if err := attempt.Check(); err != nil {
```
Comment (codingllama) [inline lib/services/local/users.go:1114]: Should we check `attempt == nil` first?

File: lib/services/local/users.go [hunk 6]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+func (s *IdentityService) GetUserRecoveryAttempts(ctx context.Context, user string) ([]types.RecoveryAttempt, error) {
+	if user == "" {
+		return nil, trace.BadParameter("missing parameter user")
+	}
+
+	startKey := backend.Key(webPrefix, usersPrefix, user, recoveryAttemptsPrefix)
+	result, err := s.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, trace.NotFound("recovery attempt record for user %q is not found", user)
```
Comment (codingllama) [inline lib/services/local/users.go:1143]: Ditto for NotFound handling, could we do without it?

File: lib/services/local/users.go [hunk 7]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+	out := make([]types.RecoveryAttempt, len(result.Items))
+	for i, item := range result.Items {
+		var a types.RecoveryAttempt
+		if err := json.Unmarshal(item.Value, &a); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		out[i] = a
+	}
+
+	sort.Sort(recoveryAttemptsChronologically(out))
```
Comment (codingllama) [inline lib/services/local/users.go:1157]: Let's document this behavior in the Identity method.

File: lib/services/local/users.go [hunk 8]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+func (s *IdentityService) DeleteUserRecoveryAttempts(ctx context.Context, user string) error {
+	if user == "" {
+		return trace.BadParameter("missing parameter user")
+	}
+
+	startKey := backend.Key(webPrefix, usersPrefix, user, recoveryAttemptsPrefix)
+	return trace.Wrap(s.DeleteRange(ctx, startKey, backend.RangeEnd(startKey)))
+}
+
+// recoveryAttemptsChronologically sorts recovery attempts by time.
```
Comment (codingllama) [inline lib/services/local/users.go:1172]: What is the sort direction?

File: lib/services/local/users.go [hunk 9]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+type recoveryAttemptsChronologically []types.RecoveryAttempt
+
+// Len returns length of a role list.
```
Comment (codingllama) [inline lib/services/local/users.go:1175]: Kill this comment, it's not adding much.

File: lib/services/local/users.go [hunk 10]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
...
+func (s recoveryAttemptsChronologically) Len() int {
+	return len(s)
+}
+
+// Less stacks latest attempts to the end of the list.
+func (s recoveryAttemptsChronologically) Less(i, j int) bool {
+	return s[i].Time.Before(s[j].Time)
+}
+
+// Swap swaps two attempts.
```
Comment (codingllama) [inline lib/services/local/users.go:1185]: Kill this comment too?

File: lib/services/identity.go [hunk 11]
```diff
@@ -207,6 +207,21 @@ type Identity interface {
...
 	GetUserTokenSecrets(ctx context.Context, tokenID string) (types.UserTokenSecrets, error)
 
+	// UpsertRecoveryCodes upserts a user's new recovery codes.
+	UpsertRecoveryCodes(ctx context.Context, user string, recovery types.RecoveryCodesV1) error
+
+	// GetRecoveryCodes gets a user's recovery codes.
+	GetRecoveryCodes(ctx context.Context, user string) (*types.RecoveryCodesV1, error)
+
+	// CreateUserRecoveryAttempt logs user recovery attempt.
+	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt types.RecoveryAttempt) error
```
Comment (codingllama) [inline lib/services/identity.go:217]: Suggestion: pass structs by pointer.

```suggestion
	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt *types.RecoveryAttempt) error
```

File: lib/services/identity.go [hunk 12]
```diff
@@ -207,6 +207,21 @@ type Identity interface {
...
+
+	// GetUserRecoveryAttempts returns user recovery attempts.
+	GetUserRecoveryAttempts(ctx context.Context, user string) ([]types.RecoveryAttempt, error)
```
Comment (codingllama) [inline lib/services/identity.go:220]: Suggestion: similarly to above, use a slice-to-pointer. Easier to reason about, IMO.

```suggestion
	GetUserRecoveryAttempts(ctx context.Context, user string) ([]*types.RecoveryAttempt, error)
```

File: lib/services/local/users.go [hunk 13]
```diff
@@ -1058,6 +1058,135 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
 	return &req, nil
 }
 
+// GetRecoveryCodes returns user's recovery codes.
+func (s *IdentityService) GetRecoveryCodes(ctx context.Context, user string) (*types.RecoveryCodesV1, error) {
```
Comment (codingllama) [inline lib/services/local/users.go:1062]: Could we add tests for the new methods here?

File: lib/web/apiserver_test.go [hunk 14]
```diff
@@ -2393,6 +2401,91 @@ func TestWebSessionsRenewAllowsOldBearerTokenToLinger(t *testing.T) {
...
+func (m *testCloudModules) Features() modules.Features {
+	return modules.Features{
+		Cloud: true, // Explicily turn on cloud feature.
+	}
+}
+
+// TestReceivingAccountRecoveryCodesWhenChangingAuthns tests for following:
+//  - Recovery codes are not returned for usernames that are not emails
+//  - Recovery codes are returned for usernames that are valid emails
+func TestReceivingAccountRecoveryCodesWhenChangingAuthenticationsWithToken(t *testing.T) {
```
Comment (codingllama) [inline lib/web/apiserver_test.go:2417]: Suggestion: Begin the test name with the method being tested, add scenarios and outcomes to the end.

```suggestion
func TestChangeUserAuthentication_recoveryCodesReturnedForCloud(t *testing.T) {
```

File: lib/web/apiserver_test.go [hunk 15]
```diff
@@ -2393,6 +2401,91 @@ func TestWebSessionsRenewAllowsOldBearerTokenToLinger(t *testing.T) {
...
+	defer modules.SetModules(defaultModules)
+	modules.SetModules(&testCloudModules{})
+
+	// Creaet a username that is not a valid email format for recovery.
+	teleUser, err := types.NewUser("invalid-name-for-recovery")
+	require.NoError(t, err)
+	env.server.Auth().CreateUser(ctx, teleUser)
+
+	// Create a reset password token and secrets.
+	token, err := env.server.Auth().CreateResetPasswordToken(ctx, auth.CreateUserTokenRequest{
```
Comment (codingllama) [inline lib/web/apiserver_test.go:2441]: Rename `token` to `resetToken`. Let's be specific about which token we are using when we can.

File: lib/web/apiserver_test.go [hunk 16]
```diff
@@ -2393,6 +2401,91 @@ func TestWebSessionsRenewAllowsOldBearerTokenToLinger(t *testing.T) {
...
+		Name: "invalid-name-for-recovery",
+	})
+	require.NoError(t, err)
+	secrets, err := env.server.Auth().RotateUserTokenSecrets(ctx, token.GetName())
+	require.NoError(t, err)
+	secondFactorToken, err := totp.GenerateCode(secrets.GetOTPKey(), env.clock.Now())
```
Comment (codingllama) [inline lib/web/apiserver_test.go:2447]:
```suggestion
	totpCode, err := totp.GenerateCode(secrets.GetOTPKey(), env.clock.Now())
```

**Review by codingllama — COMMENTED** [review]
OK, 1st pass complete.

This is __much__ better, in terms of size, compared to previous PRs, but there is still too much going on here. In the future, I would suggest separate PRs for:

* Rate limiting
* ChangeAuthentication changes (possibly multiple PRs too)
* Identity support for recovery codes (only storage)
* Recovery code implementation

(Possibly more, that's what I recall now.)

Focused PRs are easier to review, reason about, discuss and submit. It's likely we'll sort out the comments for some areas of the PR before others, but they'll have to wait along everything else to get in.

I hope the comments are useful. Thanks for the patience as well.

**Comment by codingllama** [comment]
I just noticed the merge target is not master. What is the intended merging strategy here?

**Review by codingllama — APPROVED** [review]

**Review by r0mant — APPROVED** [review]
@kimlisa Just a couple comments before you merge.

**Pull request merged**